### ISSUE TITLE
Publish fails quickly with EOF for some containers/snapshots

### ISSUE THREAD

**Issue by joeledwards** [lxc/lxd#1448]
This is happening consistently for a number of our containers. It appears that the URL which provides the mechanism for waiting on outstanding operations to complete may not have been initialized yet.

```
error: Get https://172.27.34.25:8443/1.0/operations/2cc26491-af2c-4697-aa26-5ea3204d80ba/wait: EOF
```

We experienced this with approximately half of our container snapshots, where the other half completed. Those which failed always did so consistently (retries did not help). The same is true of those which succeeded.

**Comment by tych0** [comment]
On Tue, Dec 29, 2015 at 11:19:18AM -0800, Joel Edwards wrote:
> This is happening consistently for a number of our containers. It appears that the URL which provides the mechanism for waiting on outstanding operations to complete may not have been initialized yet.
> 
> ```
> error: Get https://172.27.34.25:8443/1.0/operations/2cc26491-af2c-4697-aa26-5ea3204d80ba/wait: EOF
> ```
> 
> We experienced this with approximately half of our container snapshots, where the other half completed. Those which failed always did so consistently (retries did not help). The same is true of those which succeeded. 

I think it should only give you EOF if the goroutine handling your
request actually panic()'d; if it was just a race and you got a 404,
the client should be smart enough to render that.

Can you look in the LXD logs for the instances when this failed and
see if it printed any stack traces for a panic?

**Comment by joeledwards** [comment]
I observed this same set of INFO & WARN messages for all container, even those which did succeed.  

```
# cat lxc.log
            lxc 1451347799.543 INFO     lxc_confile - confile.c:config_idmap:1437 - read uid map: type u nsid 0 hostid 231072 range 65536
            lxc 1451347799.543 INFO     lxc_confile - confile.c:config_idmap:1437 - read uid map: type g nsid 0 hostid 231072 range 65536
            lxc 1451347799.548 WARN     lxc_cgmanager - cgmanager.c:cgm_get:994 - do_cgm_get exited with error
            lxc 1451347811.951 INFO     lxc_confile - confile.c:config_idmap:1437 - read uid map: type u nsid 0 hostid 231072 range 65536
            lxc 1451347811.951 INFO     lxc_confile - confile.c:config_idmap:1437 - read uid map: type g nsid 0 hostid 231072 range 65536
            lxc 1451347811.956 WARN     lxc_cgmanager - cgmanager.c:cgm_get:994 - do_cgm_get exited with error
```

**Comment by tych0** [comment]
On Thu, Dec 31, 2015 at 08:03:05PM -0800, Joel Edwards wrote:
> I observed this same set of INFO & WARN messages for all container, even those which did succeed.  

Sorry, I was thinking of the LXD log, not the individual container
logs. Does that log have something in it?

**Comment by joeledwards** [comment]
The file /var/log/lxd/lxd.log appears to be nothing more than an HTTP access log for the daemon. All other logs are contained within the container-specific directories.

**Comment by stgraber** [comment]
Ok, then you'll want to stop your lxd daemon and start it manually as root with:

    lxd --debug --group lxd

That should be a lot more verbose.

**Comment by jedwardsaviata** [comment]
I launched the lxd daemon manually as you described above, and it crashed when it encountered the error I described. Here is the output I recorded:

```
INFO[01-04|17:40:39] handling                                 method=GET url=/1.0 ip=@
DBUG[01-04|17:40:39] 
	{
		"type": "sync",
		"status": "Success",
		"status_code": 200,
		"metadata": {
			"api_compat": 1,
			"auth": "trusted",
			"config": {
				"core.https_address": "[::]",
				"core.trust_password": true
			},
			"environment": {
				"addresses": [
					"172.27.50.28:8443",
					"172.27.34.28:8443",
					"172.27.36.28:8443",
					"10.0.3.1:8443",
					"192.168.122.1:8443",
					"172.27.30.28:8443",
					"172.27.32.28:8443",
					"172.27.35.28:8443",
					"172.27.37.28:8443",
					"172.27.40.28:8443"
				],
				"architectures": [
					2,
					1
				],
				"driver": "lxc",
				"driver_version": "1.1.5",
				"kernel": "Linux",
				"kernel_architecture": "x86_64",
				"kernel_version": "3.19.0-39-generic",
				"server": "lxd",
				"server_pid": 22597,
				"server_version": "0.25",
				"storage": "dir",
				"storage_version": ""
			}
		},
		"operation": ""
	} 
INFO[01-04|17:40:39] handling                                 url=/1.0/images ip=@ method=POST
DBUG[01-04|17:40:39] 
	{
		"properties": {},
		"public": false,
		"source": {
			"name": "gambit-cassandra-0q-snap-2015-12-29",
			"type": "container"
		}
	} 
DBUG[01-04|17:40:39] New task operation: 786be3c1-ed9d-4eb0-b853-001e1558962b 
DBUG[01-04|17:40:39] Started task operation: 786be3c1-ed9d-4eb0-b853-001e1558962b 
DBUG[01-04|17:40:39] 
	{
		"type": "async",
		"status": "OK",
		"status_code": 100,
		"metadata": {
			"id": "786be3c1-ed9d-4eb0-b853-001e1558962b",
			"class": "task",
			"created_at": "2016-01-04T17:40:39.382146689-07:00",
			"updated_at": "2016-01-04T17:40:39.382146689-07:00",
			"status": "Running",
			"status_code": 103,
			"resources": null,
			"metadata": null,
			"may_cancel": false,
			"err": ""
		},
		"operation": "/1.0/operations/786be3c1-ed9d-4eb0-b853-001e1558962b"
	} 
INFO[01-04|17:40:39] handling                                 method=GET url=/1.0/operations/786be3c1-ed9d-4eb0-b853-001e1558962b/wait ip=@
DBUG[01-04|17:40:39] ContainerStart                           driver=storage/dir container=gambit-cassandra-0q-snap-2015-12-29
DBUG[01-04|17:40:40] ContainerStop                            driver=storage/dir container=gambit-cassandra-0q-snap-2015-12-29
panic: runtime error: slice bounds out of range

goroutine 52 [running]:
main.(*containerLXC).tarStoreFile(0xc8202f0000, 0xc820546e70, 0x3c, 0xc8205f2d80, 0xc820297b00, 0x1f, 0x7f6a08c4db28, 0xc820261130, 0x0, 0x0)
	/build/lxd-b9ly67/lxd-0.25/obj-x86_64-linux-gnu/src/github.com/lxc/lxd/lxd/container_lxc.go:2133 +0xc5b
main.(*containerLXC).Export(0xc8202f0000, 0x7f6a08c4d1e8, 0xc820152070, 0x0, 0x0)
	/build/lxd-b9ly67/lxd-0.25/obj-x86_64-linux-gnu/src/github.com/lxc/lxd/lxd/container_lxc.go:1850 +0xa2c
main.imgPostContInfo(0xc82000cc60, 0xc8200e8000, 0x0, 0x0, 0x0, 0xc82060e510, 0xc82060e4e0, 0xc82060e360, 0x27, 0x0, ...)
	/build/lxd-b9ly67/lxd-0.25/obj-x86_64-linux-gnu/src/github.com/lxc/lxd/lxd/images.go:213 +0x48a
main.imagesPost.func2(0xc8200c2630, 0x0, 0x0)
	/build/lxd-b9ly67/lxd-0.25/obj-x86_64-linux-gnu/src/github.com/lxc/lxd/lxd/images.go:760 +0x56b
main.(*operation).Run.func1(0xc8200c2630, 0xc8201708a0)
	/build/lxd-b9ly67/lxd-0.25/obj-x86_64-linux-gnu/src/github.com/lxc/lxd/lxd/operations.go:110 +0x3a
created by main.(*operation).Run
	/build/lxd-b9ly67/lxd-0.25/obj-x86_64-linux-gnu/src/github.com/lxc/lxd/lxd/operations.go:135 +0x127

goroutine 1 [semacquire]:
sync.runtime_Semacquire(0xc82031aa9c)
	/usr/lib/go/src/runtime/sema.go:43 +0x26
sync.(*WaitGroup).Wait(0xc82031aa90)
	/usr/lib/go/src/sync/waitgroup.go:126 +0xb4
main.daemon(0x0, 0x0)
	/build/lxd-b9ly67/lxd-0.25/obj-x86_64-linux-gnu/src/github.com/lxc/lxd/lxd/main.go:352 +0x534
main.run(0x0, 0x0)
	/build/lxd-b9ly67/lxd-0.25/obj-x86_64-linux-gnu/src/github.com/lxc/lxd/lxd/main.go:214 +0x86c
main.main()
	/build/lxd-b9ly67/lxd-0.25/obj-x86_64-linux-gnu/src/github.com/lxc/lxd/lxd/main.go:56 +0x24

goroutine 17 [syscall, locked to thread]:
runtime.goexit()
	/usr/lib/go/src/runtime/asm_amd64.s:1696 +0x1

goroutine 5 [syscall]:
os/signal.loop()
	/usr/lib/go/src/os/signal/signal_unix.go:22 +0x18
created by os/signal.init.1
	/usr/lib/go/src/os/signal/signal_unix.go:28 +0x37

goroutine 6 [chan receive]:
database/sql.(*DB).connectionOpener(0xc8200b8c80)
	/usr/lib/go/src/database/sql/sql.go:634 +0x45
created by database/sql.Open
	/usr/lib/go/src/database/sql/sql.go:481 +0x336

goroutine 7 [select]:
main.(*Daemon).Init.func1(0xc82000cc60)
	/build/lxd-b9ly67/lxd-0.25/obj-x86_64-linux-gnu/src/github.com/lxc/lxd/lxd/daemon.go:803 +0x136
created by main.(*Daemon).Init
	/build/lxd-b9ly67/lxd-0.25/obj-x86_64-linux-gnu/src/github.com/lxc/lxd/lxd/daemon.go:813 +0x1776

goroutine 66 [chan receive]:
main.(*operation).WaitFinal(0xc8200c2630, 0xffffffffffffffff, 0xc8200c2630, 0x0, 0x0)
	/build/lxd-b9ly67/lxd-0.25/obj-x86_64-linux-gnu/src/github.com/lxc/lxd/lxd/operations.go:291 +0xa1
main.operationAPIWaitGet(0xc82000cc60, 0xc8200e82a0, 0x0, 0x0)
	/build/lxd-b9ly67/lxd-0.25/obj-x86_64-linux-gnu/src/github.com/lxc/lxd/lxd/operations.go:514 +0x1fc
main.(*Daemon).createCmd.func1(0x7f6a05202518, 0xc8200c2210, 0xc8200e82a0)
	/build/lxd-b9ly67/lxd-0.25/obj-x86_64-linux-gnu/src/github.com/lxc/lxd/lxd/daemon.go:285 +0x579
net/http.HandlerFunc.ServeHTTP(0xc8202daa50, 0x7f6a05202518, 0xc8200c2210, 0xc8200e82a0)
	/usr/lib/go/src/net/http/server.go:1422 +0x3a
github.com/gorilla/mux.(*Router).ServeHTTP(0xc82029d5e0, 0x7f6a05202518, 0xc8200c2210, 0xc8200e82a0)
	/build/lxd-b9ly67/lxd-0.25/obj-x86_64-linux-gnu/src/github.com/gorilla/mux/mux.go:100 +0x29e
net/http.serverHandler.ServeHTTP(0xc82011a600, 0x7f6a05202518, 0xc8200c2210, 0xc8200e82a0)
	/usr/lib/go/src/net/http/server.go:1862 +0x19e
net/http.(*conn).serve(0xc82052aa50)
	/usr/lib/go/src/net/http/server.go:1361 +0xbee
created by net/http.(*Server).Serve
	/usr/lib/go/src/net/http/server.go:1910 +0x3f6

goroutine 9 [select]:
main.deviceTaskScheduler(0xc82000cc60)
	/build/lxd-b9ly67/lxd-0.25/obj-x86_64-linux-gnu/src/github.com/lxc/lxd/lxd/devices.go:307 +0x684
created by main.(*Daemon).Init
	/build/lxd-b9ly67/lxd-0.25/obj-x86_64-linux-gnu/src/github.com/lxc/lxd/lxd/daemon.go:838 +0x19bd

goroutine 18 [syscall]:
syscall.Syscall(0x0, 0x8, 0xc8201a2000, 0x1000, 0x0, 0x5c6675, 0x9c85a0)
	/usr/lib/go/src/syscall/asm_linux_amd64.s:18 +0x5
syscall.read(0x8, 0xc8201a2000, 0x1000, 0x1000, 0x1000, 0x0, 0x0)
	/usr/lib/go/src/syscall/zsyscall_linux_amd64.go:783 +0x5f
syscall.Read(0x8, 0xc8201a2000, 0x1000, 0x1000, 0x1000, 0x0, 0x0)
	/usr/lib/go/src/syscall/syscall_unix.go:160 +0x4d
main.deviceMonitorProcessors.func1(0x800, 0x8, 0xc820168000)
	/build/lxd-b9ly67/lxd-0.25/obj-x86_64-linux-gnu/src/github.com/lxc/lxd/lxd/devices.go:68 +0xa6
created by main.deviceMonitorProcessors
	/build/lxd-b9ly67/lxd-0.25/obj-x86_64-linux-gnu/src/github.com/lxc/lxd/lxd/devices.go:102 +0x200

goroutine 12 [chan receive]:
main.daemon.func2(0xc82000cc60, 0xc82031aa80, 0xc82031aa90)
	/build/lxd-b9ly67/lxd-0.25/obj-x86_64-linux-gnu/src/github.com/lxc/lxd/lxd/main.go:317 +0x16e
created by main.daemon
	/build/lxd-b9ly67/lxd-0.25/obj-x86_64-linux-gnu/src/github.com/lxc/lxd/lxd/main.go:326 +0x4ba

goroutine 13 [chan receive]:
main.daemon.func3(0xc82000cc60, 0xc82031aa80, 0xc82031aa90)
	/build/lxd-b9ly67/lxd-0.25/obj-x86_64-linux-gnu/src/github.com/lxc/lxd/lxd/main.go:329 +0x41
created by main.daemon
	/build/lxd-b9ly67/lxd-0.25/obj-x86_64-linux-gnu/src/github.com/lxc/lxd/lxd/main.go:338 +0x4f0

goroutine 14 [chan receive]:
main.daemon.func4(0xc82031aa80, 0xc82000cc60, 0xc82031aa90)
	/build/lxd-b9ly67/lxd-0.25/obj-x86_64-linux-gnu/src/github.com/lxc/lxd/lxd/main.go:345 +0x354
created by main.daemon
	/build/lxd-b9ly67/lxd-0.25/obj-x86_64-linux-gnu/src/github.com/lxc/lxd/lxd/main.go:350 +0x526

goroutine 15 [select, locked to thread]:
runtime.gopark(0xd11388, 0xc820324728, 0xb4ca50, 0x6, 0x18, 0x2)
	/usr/lib/go/src/runtime/proc.go:185 +0x163
runtime.selectgoImpl(0xc820324728, 0x0, 0x18)
	/usr/lib/go/src/runtime/select.go:392 +0xa64
runtime.selectgo(0xc820324728)
	/usr/lib/go/src/runtime/select.go:212 +0x12
runtime.ensureSigM.func1()
	/usr/lib/go/src/runtime/signal1_unix.go:227 +0x353
runtime.goexit()
	/usr/lib/go/src/runtime/asm_amd64.s:1696 +0x1

goroutine 16 [IO wait]:
net.runtime_pollWait(0x7f6a08c4ed30, 0x72, 0xc82000e1e0)
	/usr/lib/go/src/runtime/netpoll.go:157 +0x60
net.(*pollDesc).Wait(0xc82031c290, 0x72, 0x0, 0x0)
	/usr/lib/go/src/net/fd_poll_runtime.go:73 +0x3a
net.(*pollDesc).WaitRead(0xc82031c290, 0x0, 0x0)
	/usr/lib/go/src/net/fd_poll_runtime.go:78 +0x36
net.(*netFD).accept(0xc82031c230, 0x0, 0x7f6a08c53488, 0xc82054c160)
	/usr/lib/go/src/net/fd_unix.go:408 +0x27c
net.(*UnixListener).AcceptUnix(0xc8203010c0, 0xc820054e00, 0x0, 0x0)
	/usr/lib/go/src/net/unixsock_posix.go:304 +0x53
net.(*UnixListener).Accept(0xc8203010c0, 0x0, 0x0, 0x0, 0x0)
	/usr/lib/go/src/net/unixsock_posix.go:314 +0x41
net/http.(*Server).Serve(0xc82011a600, 0x7f6a08c53560, 0xc8203010c0, 0x0, 0x0)
	/usr/lib/go/src/net/http/server.go:1887 +0xb3
net/http.Serve(0x7f6a08c53560, 0xc8203010c0, 0x7f6a041bc000, 0xc82029d5e0, 0x0, 0x0)
	/usr/lib/go/src/net/http/server.go:1765 +0x85
main.(*Daemon).Init.func5.1(0x0, 0x0)
	/build/lxd-b9ly67/lxd-0.25/obj-x86_64-linux-gnu/src/github.com/lxc/lxd/lxd/daemon.go:978 +0x7d
gopkg.in/tomb%2ev2.(*Tomb).run(0xc82000ccf8, 0xc82033c150)
	/build/lxd-b9ly67/lxd-0.25/obj-x86_64-linux-gnu/src/gopkg.in/tomb.v2/tomb.go:153 +0x21
created by gopkg.in/tomb%2ev2.(*Tomb).Go
	/build/lxd-b9ly67/lxd-0.25/obj-x86_64-linux-gnu/src/gopkg.in/tomb.v2/tomb.go:149 +0x131

goroutine 50 [IO wait]:
net.runtime_pollWait(0x7f6a08c4ec70, 0x72, 0xc82000e1e0)
	/usr/lib/go/src/runtime/netpoll.go:157 +0x60
net.(*pollDesc).Wait(0xc82031c300, 0x72, 0x0, 0x0)
	/usr/lib/go/src/net/fd_poll_runtime.go:73 +0x3a
net.(*pollDesc).WaitRead(0xc82031c300, 0x0, 0x0)
	/usr/lib/go/src/net/fd_poll_runtime.go:78 +0x36
net.(*netFD).accept(0xc82031c2a0, 0x0, 0x7f6a08c53488, 0xc82036a000)
	/usr/lib/go/src/net/fd_unix.go:408 +0x27c
net.(*TCPListener).AcceptTCP(0xc820048638, 0x4, 0x0, 0x0)
	/usr/lib/go/src/net/tcpsock_posix.go:254 +0x4d
net.(*TCPListener).Accept(0xc820048638, 0x0, 0x0, 0x0, 0x0)
	/usr/lib/go/src/net/tcpsock_posix.go:264 +0x3d
crypto/tls.(*listener).Accept(0xc820301180, 0x0, 0x0, 0x0, 0x0)
	/usr/lib/go/src/crypto/tls/tls.go:46 +0x60
net/http.(*Server).Serve(0xc82034a060, 0x7f6a08c53670, 0xc820301180, 0x0, 0x0)
	/usr/lib/go/src/net/http/server.go:1887 +0xb3
net/http.Serve(0x7f6a08c53670, 0xc820301180, 0x7f6a041bc000, 0xc82029d5e0, 0x0, 0x0)
	/usr/lib/go/src/net/http/server.go:1765 +0x85
main.(*Daemon).Init.func5.1(0x0, 0x0)
	/build/lxd-b9ly67/lxd-0.25/obj-x86_64-linux-gnu/src/github.com/lxc/lxd/lxd/daemon.go:978 +0x7d
gopkg.in/tomb%2ev2.(*Tomb).run(0xc82000ccf8, 0xc82033c2a0)
	/build/lxd-b9ly67/lxd-0.25/obj-x86_64-linux-gnu/src/gopkg.in/tomb.v2/tomb.go:153 +0x21
created by gopkg.in/tomb%2ev2.(*Tomb).Go
	/build/lxd-b9ly67/lxd-0.25/obj-x86_64-linux-gnu/src/gopkg.in/tomb.v2/tomb.go:149 +0x131

goroutine 51 [IO wait]:
net.runtime_pollWait(0x7f6a08c4edf0, 0x72, 0xc82000e1e0)
	/usr/lib/go/src/runtime/netpoll.go:157 +0x60
net.(*pollDesc).Wait(0xc8200d7720, 0x72, 0x0, 0x0)
	/usr/lib/go/src/net/fd_poll_runtime.go:73 +0x3a
net.(*pollDesc).WaitRead(0xc8200d7720, 0x0, 0x0)
	/usr/lib/go/src/net/fd_poll_runtime.go:78 +0x36
net.(*netFD).accept(0xc8200d76c0, 0x0, 0x7f6a08c53488, 0xc820301a00)
	/usr/lib/go/src/net/fd_unix.go:408 +0x27c
net.(*UnixListener).AcceptUnix(0xc8200fb200, 0xc8200d1e20, 0x0, 0x0)
	/usr/lib/go/src/net/unixsock_posix.go:304 +0x53
net.(*UnixListener).Accept(0xc8200fb200, 0x0, 0x0, 0x0, 0x0)
	/usr/lib/go/src/net/unixsock_posix.go:314 +0x41
net/http.(*Server).Serve(0xc82033e780, 0x7f6a08c53560, 0xc8200fb200, 0x0, 0x0)
	/usr/lib/go/src/net/http/server.go:1887 +0xb3
main.(*Daemon).Init.func5.2(0x0, 0x0)
	/build/lxd-b9ly67/lxd-0.25/obj-x86_64-linux-gnu/src/github.com/lxc/lxd/lxd/daemon.go:983 +0x7b
gopkg.in/tomb%2ev2.(*Tomb).run(0xc82000ccf8, 0xc82031b0d0)
	/build/lxd-b9ly67/lxd-0.25/obj-x86_64-linux-gnu/src/gopkg.in/tomb.v2/tomb.go:153 +0x21
created by gopkg.in/tomb%2ev2.(*Tomb).Go
	/build/lxd-b9ly67/lxd-0.25/obj-x86_64-linux-gnu/src/gopkg.in/tomb.v2/tomb.go:149 +0x131
```

**Comment by stgraber** [comment]
Thanks, that stacktrace is perfect for us to figure out what happened.
